LEAN-2650: Introduce new config API

mbetamony opened the pull request.

mbetamony commented:
Hello @asouqi, this is not yet ready for review but I'm adding you as a reviewer to keep you up to date on what I'm working on

mbetamony commented:
@hhusban and @mnatsheh I will add you as reviewers to keep you in the loop of what I'm working on and if you would prefer to do things some other way

For now I've done the following: 

Introduce new routes, schema, controller and service for the new Config-api

The main logic happens in the ConfigService:

If a specific file name is provided, the service directly reads and returns the content of the file, or if no file name is given, it returns the aggregated data of all the JSON files in the given directory.
Also, if the file is read once, its saved in memory so we don't have to read it again from desk again (now we can implement some caching options and keep a limit on how much memory we should use) 
And if only one file is selected, we return the file (either from memory or desk) and then read other .json files in the same directory into memory for future usages.

the routes are as follows `(GET)`: 
```
config/shared/:fileName?
config/locales/:fileName?
config/styles/:fileName?
```
Next step is to clean the code up a bit, add unit testing, and add documentation (I'm currently looking at swagger or/and tsoa) 
Please let me know if I should consider different approaches, thank you.

asouqi commented:
Can we add also endpoint to return templates model by it's id, like this for example:
```
config/shared/templates
parameters: [MPTemplate:Id#1, ..., MPTemplate:Id#N]
```
I suggested that, to avoid bringing the whole file `templates-v2.json` with size (9.81 MB) in FE, so we could request just the models we need.

CC: @mbetamony @hhusban

mbetamony commented:
> Can we add also endpoint to return templates model by it's id, like this for example:
> 
> ```
> config/shared/templates
> parameters: [MPTemplate:Id#1, ..., MPTemplate:Id#N]
> ```
> 
> I suggested that, to avoid bringing the whole file `templates-v2.json` with size (9.81 MB) in FE, so we could request just the models we need.
> 
> also if we could do that for `bundles.json` that will be great.
> 
> CC: @mbetamony @hhusban

Yes we can, I'll implement that. Do we need that for other files as well? bundles.json for example (its a similar implementation but whether we should or shouldn't expose an api route for that)

asouqi commented:
>  Do we need that for other files as well? bundles.json

yes for sure

asouqi commented:
>bundles
section-categories
templates-v2

for `section-categories` I don't think will use it as read all the list. and in the request parameters will pass array of the Model ids we need?

mbetamony commented:
> > bundles
> > section-categories
> > templates-v2
> 
> for `section-categories` I don't think will use it as read all the list. and in the request parameters will pass array of the Model ids we need?

I still don't have the params figured out, I'll let you know once its decided

hhusban commented:
> I still don't have the params figured out, I'll let you know once its decided

You can keep the uri as following,
`${this.basePath}/shared/:fileName?'
 and use req.query to communicate ids you want filter with
 
a sample url for this would be
/config/shared/templates?ids=tem1,tem2,...,temN

mbetamony commented:
I went with the approach @hhusban  mentioned, you can now retrieve either a full file, an array of properties or a just a single property. 
 /config/shared/templates?ids=tem1,tem2,...,temN

mbetamony commented:
@asouqi @mbartenev-atypon 
I wanted to ask, are you going to cache things on your side? 
take the section-categories drop-down as an example, are you going to send a request to the API for each reload or are you going to send a request once and save the response in memory? if that's the case I think we don't need to cache anything here. 
please let me know, thanks.

mbartenev-atypon commented:
The routes look ok to me.

mbartenev-atypon commented:
> @asouqi @mbartenev-atypon I wanted to ask, are you going to cache things on your side? take the section-categories drop-down as an example, are you going to send a request to the API for each reload or are you going to send a request once and save the response in memory? if that's the case I don't think we need to cache anything here. please let me know, thanks.

We are not planning to cache anything from this config api at the moment. It's a good idea to cache it on FE but we might do it later.

mbetamony commented:
@hhusban and @mnatsheh 
I moved this to in-review, once reviewed and you are OK with the implementation I can proceed with unit-testing and documentation.

mnatsheh commented:
Based on discussions with various people, let's make the following changes:

- Let's remove the `/config` prefix from the endpoints. The endpoints I suggest are:
  - `/template/<ID>`
  - `/bundle/<ID>`
  - `/csl/style/<ID>`
  - `/csl/locale/<ID>`
  - `/secion-categories` (more on this later)
(alternativally, we could pass <ID> as a query param instead of a path param, I'm fine with either)

- We don't need to copy the manuscripts-data build logic and DBs here. We only want a handful of templates anyway. The templates we want to preserve are:
  - MPManuscriptTemplate:www-zotero-org-styles-nature-genetics-Nature-Genetics-Journal-Publication
  - MPManuscriptTemplate:www-zotero-org-styles-bmc-biology-BMC-Biology-Journal-Publication-Editorial
  - MPManuscriptTemplate:MPManuscriptTemplate:MPManuscriptTemplate:MPBundle:www-zotero-org-styles-harvard1-cambridge_university_press_journal_publication-MPBundle:www-zotero-org-styles-chicago-author-date-health_economics,_policy_and_law-MPBundle:www-zotero-org-styles-chicago-author-date-review_article_health_economics,_policy_and_law
  - MPManuscriptTemplate:www-zotero-org-styles-nature-genetics-Nature-Genetics-Journal-Publication-Article
For each of them, we should keep the templates, all its parents, its bundle, and citation style. Everything else should be removed.

Regarding the section categories, @asouqi pointed out to me that we use them in manuscripts-transform in relation to the new section groups we added. To avoid adding a dependency from transform to api, we should move the definition of these 3 section categories into transform (let's call them the "core" section categories), while the new api endpoint will return the "extra" section categories. The UI app will be responsible for combining the 2 lists for its purposes.

I also discussed some internal changes to how this work.

mbetamony commented:
I removed the old data, and generated the new data based on the templates we need.
I'm also  not sure if we need the `metadata.json` for styles, I'll delete it unless we need it. 
Please let me know if this is sufficient for how we need the data.
I kept `locales.json` the same, and removed the 3 core sections from `section-categories.json`
Here are the some snippets of the scripts I used to generate the data:

Templates
```
const templateIDs = [
    'MPManuscriptTemplate:www-zotero-org-styles-nature-genetics-Nature-Genetics-Journal-Publication',
    'MPManuscriptTemplate:www-zotero-org-styles-bmc-biology-BMC-Biology-Journal-Publication-Editorial',
    'MPManuscriptTemplate:MPManuscriptTemplate:MPManuscriptTemplate:MPBundle:www-zotero-org-styles-harvard1-cambridge_university_press_journal_publication-MPBundle:www-zotero-org-styles-chicago-author-date-health_economics,_policy_and_law-MPBundle:www-zotero-org-styles-chicago-author-date-review_article_health_economics,_policy_and_law',
    'MPManuscriptTemplate:www-zotero-org-styles-nature-genetics-Nature-Genetics-Journal-Publication-Article'
];

templateIDs.forEach(id => {
    let currentItem = data.find(d => d._id === id);
    if(currentItem){
        keep.add(currentItem._id);
        while (currentItem && currentItem.parent) {
            let parentItem = data.find(d => d._id === currentItem.parent || d.title === currentItem.parent);
            if (parentItem) {
                keep.add(parentItem._id);
                currentItem = parentItem;
            } else {
                currentItem = null;
            }
        }
    }
});

let filteredData = data.filter(d => keep.has(d._id));
// Remove the "mandatorySectionRequirements" attribute
filteredData = filteredData.map(item => {
    delete item.mandatorySectionRequirements;
    return item;
});

fs.writeFileSync(path.join(root, '/config/templates.json'), JSON.stringify(filteredData, null, 2));
 ```
Bundles
```
const filteredTemplates = JSON.parse(fs.readFileSync(filteredTemplatesPath, 'utf8'));
const bundles = JSON.parse(fs.readFileSync(bundlesPath, 'utf8'));

// Get unique bundle values from the filtered templates
const usedBundles = [...new Set(filteredTemplates.map(item => item.bundle))];

// Filter bundles.json
const filteredBundles = bundles.filter(bundle => usedBundles.includes(bundle._id));

fs.writeFileSync(path.join(root, '/config/bundles.json'), JSON.stringify(filteredBundles, null, 2));
 ```
Styles, this is almost identical to how we managed the old styles, but this generates all of them into one file and gets the styles we need from `bundles.json`
``` 
const bundlesPath = path.join(__dirname, '/../../dist/config/bundles.json');
let bundlesData = JSON.parse(fs.readFileSync(bundlesPath, 'utf-8'));

let neededCslStyles = [];
bundlesData.forEach(bundle => {
    if (bundle.csl && bundle.csl.cslIdentifier) {
        neededCslStyles.push(bundle.csl.cslIdentifier);
    }
});

console.log("Needed CSL Styles:", neededCslStyles);

const outputDir = __dirname + '/../../dist/config/';
if (fs.existsSync(outputDir)) {
    fs.rmdirSync(outputDir, { recursive: true });
}
fs.mkdirSync(outputDir);

const styles = {};
const items = {};

const stylesDir = __dirname + '/../csl/styles';
const stylePaths = globby.sync([stylesDir + '/*.csl', stylesDir + '/dependent/*.csl']);

for (const stylePath of stylePaths) {
    const xml = fs.readFileSync(stylePath, 'utf-8');
    const doc = sync(xml, {});
    const info = evaluateXPathToFirstNode('style/info', doc);
    const uri = evaluateXPathToString('link[@rel="self"]/@href', info);
    if (!uri) throw new Error('Self identifier not found');

    // Skip the CSL style if it is not needed
    if (!neededCslStyles.includes(uri)) {
        continue;
    }

    console.log("Matched CSL Style:", uri); 

    styles[uri] = xml;

    items[uri] = {
        issn: evaluateXPathToString('issn', info) || undefined,
        eissn: evaluateXPathToString('eissn', info) || undefined,
        title: evaluateXPathToString('title', info) || undefined,
        short: evaluateXPathToString('title-short', info) || undefined,
        category: evaluateXPathToArray('array{category/@field}', info).map((node) => node.nodeValue),
    };
}

fs.writeFileSync(outputDir + `/styles.json`, JSON.stringify(styles));
fs.writeFileSync(outputDir + `/metadata.json`, JSON.stringify(items));
```

mbetamony commented:
- Updated the data
- Updated the routes (I added an 's' for plural since I used queries in the routes)
- Added a new environment variable `CONFIG_PATH=./config`
- Updated the dockerfile to copy the data` COPY ./config ./config`

For caching:
We load all files at startup, then if we request anything it will be cached (I can still play around with this and cache all objects in the file) 

Please let me know if any more updates are needed
@mnatsheh , @hhusban

mnatsheh commented on config/bundles.json:
```diff
@@ -0,0 +1,16 @@
+    "csl": {
+      "_id": "MPCitationStyle:www-zotero-org-styles-american-medical-association"
+    }
+  },
```
@asouqi let's switch to using the `_id` value.

mnatsheh commented on config/templates.json:
```diff
@@ -0,0 +1,34 @@
+    "title": "Nature Genetics Journal Publication",
+    "createdAt": 1433095330.89114,
+    "updatedAt": 1433095363.40453
+  },
```
We will lose the quality report functionality from the FE, but I don't think requirements will break (i.e. the manuscript will still open).

mbetamony commented on src/Controller/V2/Config/ConfigSchema.ts:
```diff
@@ -0,0 +1,45 @@
+
+export const configSchema: Joi.SchemaMap = {
+  headers: jsonHeadersSchema.headers,
+}
```
Shouldn't we add a query object here with id? 
And shouldn't we also make all ids inside different queries required? 
@mnatsheh

mbetamony replied:
@mnatsheh The id query to the `configSchema` has been added by the this commit [Update schemas headers](https://github.com/Atypon-OpenSource/manuscripts-api/pull/62/commits/88fe0643636456bad906578f1bd70a4b9c392ecf), which was added  after this comment. 
I'll make the id required and add the regex.

mnatsheh replied:
Got it. Thanks.

mbetamony commented:
- Schemas have been updated to use correct headers, and documentation has been added under host/api-docs. 
- Data has been corrected in section-categories.json
This is ready for review. 
CC: @mnatsheh @hhusban

mbetamony commented:
The latest commit includes truncating bundle.csl._id
@asouqi , please let me know when you update the [article-editor](https://github.com/Atypon-OpenSource/manuscripts-article-editor/pull/241) so I can publish this

asouqi commented:
> The latest commit includes truncating bundle.csl._id
> @asouqi , please let me know when you update the [article-editor](https://github.com/Atypon-OpenSource/manuscripts-article-editor/pull/241) so I can publish this

@mnatsheh are you ok with the changes in `article-editor` [PR](https://github.com/Atypon-OpenSource/manuscripts-article-editor/pull/241)? so we can publish current updates

mbetamony commented:
@mnatsheh Can we decrease the coverage rate for tests here? I don't think there's any need to unit test anything in controller or route.

mnatsheh commented:
Yes, you can decrease the coverage.

On Sun, Aug 13, 2023, 9:51 AM Mohammad Betamony ***@***.***>
wrote:

> @mnatsheh <https://github.com/mnatsheh> Can we decrease the coverage rate
> for tests here? I don't think there's any need to unit test anything in
> controller or route.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Atypon-OpenSource/manuscripts-api/pull/62#issuecomment-1676410773>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGHKKSQOSYOZWY7S4I53L7TXVEAYJANCNFSM6AAAAAA2M2HWJ4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
